### DISCUSSION TITLE
integrating players with different strategies?

### DISCUSSION THREAD

**Discussion by luoziqing99** [deepmind/open_spiel#1145] [General]
is there an agent/player class? If we want to use players with different strategies to play a certain game (ex. Khun poker), is there a base agent class that we can inherit from?

**Comment by lanctot** [comment]
Yes. 

There are two notions of these in OpenSpiel: bots and agents.

1. There are **bots**, the basic API is outlined in [spiel_bots.h](https://github.com/google-deepmind/open_spiel/blob/master/open_spiel/spiel_bots.h) and several examples [here](https://github.com/google-deepmind/open_spiel/tree/master/open_spiel/bots) and in [python/bots](https://github.com/google-deepmind/open_spiel/tree/master/open_spiel/python/bots).

2. There's a basic RL Agent API outlined in [python/rl_agent.py](https://github.com/google-deepmind/open_spiel/blob/master/open_spiel/python/rl_agent.py), and several examples in [python/algorithms](https://github.com/google-deepmind/open_spiel/tree/master/open_spiel/python/algorithms), such as e.g. mcts_agent.py and dqn_agent.py.

They differ in their intended use. Bots expose a few functions but mainly `step(State) -> Action`. The bots get the full state and return an action. These are appropriate for agents that need the full state (often for planning.. think of a minimax or MCTS agent). An RL agent exposes a similar function `step(timestep, is_evaluation=False)`. The timestep object resembles similar ones in other widely-used RL APIs like e.g. gym and DM_env. They contain observations that are often used to train the agents.

Which one you use depends on the context, but it often comes down to the simple advice of "Do you intend to do some RL? If yes, use the Agent class. If not, use a bot" but it really depends on what you want to do, so please follow-up if it's not clear.
